### ISSUE TITLE
Question re: SSR

### ISSUE THREAD

**Issue by StevenLangbroek** [redux-observable/redux-observable#47]
If I use React Router's `match` function, and trigger actions from all matched components, is there a good way to coordinate when all these actions (Observables) are `completed`? This would be essential for SSR, and is normally coordinated with Promises and `Promise.all`.

**Comment by istarkov** [comment]
You could add a redux-observable processor with server process like 

https://jsfiddle.net/u1gay0qj/

```javascript
Rx.Observable.combineLatest(
	// use actions$.ofType('MY_ACTION')
  actions$.filter(a => a.type === 'MY_ACTION').take(1),
  actions$.filter(a => a.type === 'MY_OTHER_ACTION').take(1),
  // all other actions
)
.delay(0)
.do(() => console.log('HERE YOU CAN SAVE STATE'))
```

**Comment by istarkov** [comment]
In my case as I use for data fetch standalone process, like this https://github.com/istarkov/redux-observable-process-fetch/blob/master/src/createFetchProcess.js
I just could count `LOADING_START`, `LOADING_END` (`_CANCEL`) actions, and when 
`count(LOADING_START) === count(LOADING_END)` I could save state

**Comment by jayphelps** [comment]
@StevenLangbroek To be honest, neither myself or @blesh have much experience with React SSR since none of the apps we work on need it.

This is a great question that we _should_ provide guidance on..we will either need to familiarize ourselves with React SSR or defer to the experiences of some of our community members. As @istarkov mentioned, we recently introduced the ability to define "process managers" that are very similar conceptually to redux-saga (which are actually process managers too, not _true_ sagas since they maintain state) so I imagine there will be existing solutions we can learn from. e.g. https://github.com/yelouafi/redux-saga/blob/master/examples/real-world/server.js but you'll quickly see that redux-saga is not redux-observable--for one, there's no first-class concept of tasks.

We're working on a major overhaul of docs, I'll add this to the list of items to explore. If anyone has ideas, feel free to share them here.

**Comment by StevenLangbroek** [comment]
@jayphelps thanks for the reply. Conceptually speaking, the process is relatively straight forward:

- match routes to an array of components
- pick static methods off of them
- map over them and call them with `dispatch`, `params`, `query` and what else you need to be able dispatch actions for the routes
- coordinate their total completion somehow, and carry on with rendering

I'm less experienced with rx / observables, so the 2 main questions I'd have to be able to chime in is:

- what sort of construct would you gather the observables in? an observable of observables? i'm out of my depth here... in the case of redux-thunk, you map your promises into an array, and wait for them to resolve with `Promise.all`. 
- how do you signal a data dependency to be "resolved"? 

I think my initial question was maybe colored too much by my experience with thunks / Promises. Maybe there is a more natural fit in the world of Observables that I'm not aware of... Maybe we can gather thoughts / discussion in this thread (or elsewhere if it's already happening)...

**Comment by StevenLangbroek** [comment]
@istarkov, looking at your snippet above, I'm imagining something like this:

```js
import React, { Component } from 'react';
import { fetchPost, RECEIVE_POST } from 'reducers/blog';

export default BlogPost extends Component {

  static fetch({ dispatch, params }) {
    return {
      complete: RECEIVE_POST,
      observable: dispatch(fetchPost(params.postId))
    };
  }

  render() {
    /* ... */
  }
}
```

Where the complete properties are gathered, and waited on with `combineLatest` like in your snippet. I'm still going through the code in your fetch-process to understand it better, please correct me if I'm misunderstanding something.

**Comment by StevenLangbroek** [comment]
```
export default (req, res, next) => {
  match({
    history,
    routes,
    location,
  }, (error, redirectLocation, renderProps) => {
    const { components } = renderProps;

    const fetches = components
      .filter(component => component) // filter null components
      .map(component => component.fetch) // pick off fetch methods
      .filter(fetcher => fetcher) // filter copnonents without fetch
      .map(fetcher => fetcher({ dispatch, actions, params, query })); // trigger observables

    const completions = fetches.map({ complete }) => complete)

    // TODO: somehow coordinate when these completion actions have all fired
    // TODO: then call next(); and actually start rendering out components.
  }
}
```

**Comment by istarkov** [comment]
As I understand your case is something like [redial](https://github.com/markdalgleish/redial)
So somewhere in you code you have
```
  match({ routes, history }, (error, redirectLocation, renderProps) => {
    // Get array of route handler components:
    const { components } = renderProps;   
```

And now you can iterate over `components` collection and read static `fetch` function.

As a simple example you could just return in `fetch` something like 
```
        Observable.ajax({
          type: 'GET',
          url: `blbla${param}blabalUrl`
          crossDomain: true,
        })
        .map(ajax => {type:'BLBLA_TYPE', payload: ajax.response}),
```

Then you could just
```
Observable.from(components)
  .filter(c => !!c.fetch)
  .mergeMap({ fetch }  => fetch(params))
  .subscribe({
    next: v => dispatch(v),
    complete: () => console.log('Data fetch done')
  })
```

You can play here
https://jsfiddle.net/krq65mcb/

So you can do SSR even without redux-observable

**Comment by jayphelps** [comment]
Note: haven't read the above comments yet, documenting what I've found so far, not necessarily directing this to a particular person.

So far I've found that redux-saga apparently supports a special `END` action that tells some of their operators to do special behavior. https://github.com/yelouafi/redux-saga/tree/master/docs/api#takepattern usually to terminate them, after any pending forks complete.

Their example [dispatches this here](https://github.com/yelouafi/redux-saga/commit/0dc376107d24537028bd1c66e4fce81c223f2c23#diff-0bfe8d7a830f65627e298cad73547973R14).

One could put together a similar mechanism, though as of now you have to dispatch the `END` action when you're really done.

```js
const fetchUser = id => ({ type: FETCH_USER, payload: id });

const fetchUserManager = (actions, store) =>
  actions.ofType(FETCH_USER)
    .switchMap(action =>
      getJSON(`/api/users/${action.payload}`)
        .mergeMap(response => Observable.of(
          { type: FETCH_USER_FULFILLED, payload: response },
          { type: 'END' }
        ))
    );

const endListeners = new Subject();

const listenForEnd = () =>
  new Promise((resolve, reject) => {
    endListeners.next(
      (actions, store) =>
        actions.ofType('END')
          .do({ next: resolve, error: reject })
          .filter(_ => false)
    );
  });

const endListenersManager = (actions, store) =>
  endListeners.mergeMap(manager => {
    return manager(actions, store)
  });

const rootManager = combineDelegators(
  fetchUserManager,
  endListenersManager
);

const store = createStore(rootReducer,
  applyMiddleware(
    reduxObservable(rootManager)
  )
);

// server.js

match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
  if (renderProps && renderProps.components) {
    const rootComp = <Root store={store} routes={routes} history={createMemoryHistory()} renderProps={renderProps} type="server" />;

    listenForEnd().then(() => {
      res.status(200).send(
        layout(
          renderToString(rootComp),
          JSON.stringify(store.getState())
        )
      )
    });

    renderToString(rootComp);
});

// UserPage.jsx

const mapStateToProps = (state, ownProps) => ({
  userId: ownProps.params.userId,
  user: selectUserById(state, ownProps.params.userId)
});

const mapDispatchToProps = {
  fetchUser
};

@connect(mapStateToProps, mapDispatchToProps)
class UserPage extends Component {
  componentWillMount() {
    this.props.fetchUser(this.props.userId);
  }

  render() {
    const { user } = this.props;

    return (
      <div>
        {user ? <User {...user} /> : 'Loading...'}
      </div>
    );
  }
}
```

**Comment by jayphelps** [comment]
Here's a demo of adding the ability to start and stop the process manager and using that to decide when we're "done" since it will act just like redux-saga now: when you `stop()`, no further actions will processed but the store will receive any actions you dispatch.

http://jsbin.com/hibafi/edit?js,output

```js
const managerMiddleware = reduxObservable(fetchUserManager);

const store = createStore(rootReducer, applyMiddleware(managerMiddleware));

function fakeServerRequest() {
  managerMiddleware.start().then(() => {
    console.log('complete with state:', store.getState());
  });

  store.dispatch({ type: 'FETCH_USER' });

  managerMiddleware.stop(); 
}

setInterval(fakeServerRequest, 1000);
```
I can already think of several issues with this approach though:

* Concurrent requests will collide actions with each other. I'm not sure how redux-saga avoids this either?
* You cannot compose process managers async, since any actions you emit async won't be processed, only dispatched to the store--which is the point of this, but I think the implementation should be reconsidered cause loss of composability sucks and is confusing.

**Comment by jayphelps** [comment]
To clarify what I mean by composing async process managers sequentially, I mean that you wouldn't be able to do this:

```js
const fetchFirstManager = (actions, store) =>
  actions.ofType(FETCH_FIRST)
    .switchMap(action =>
      ajax('/first')
        .map(response => ({ type: FETCH_SECOND, payload: response }))
    );

const fetchSecondManager = (actions, store) =>
  actions.ofType(FETCH_SECOND)
    .switchMap(action =>
      ajax('/second')
        .map(response => ({ type: OK_IM_DONE, payload: response }))
    );
```

More commonly, you'll start with one action, then emit a series of them to delegate to other process managers: http://jsbin.com/yetolo/edit?js,output

**Comment by BerkeleyTrue** [comment]
Hello Everyone,

I'm the creator of Redux-Epic, which is a redux library in similar fashion to redux-saga but built using RxJs with first class support for SSR.

@jayphelps You are on the right track.

Redux-epic's epic middleware is sort of a [pseudo observable](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/create-epic.js#L78), you can subscribe and dispose it the same way you can an observable, but it only let's you know when it has completed (no Next or Error track). (This was created to be used internally and not a user facing feature.)

An epic has completed when all the observables returned by the sagas used to create it also complete. This is how we detect when all of our data has been fetched.

These methods above are used by a custom [`renderToString`](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/render-to-string.js#L12) function which manages the lifecycle of the epic middleware. 
* A first initial render is done to activate the epic middleware, the users sagas, and let actions be dispatched by the app.
* The epic middleware is ended (which triggers the actions$ to also call `onCompleted`)
* Then we wait for the pseudo observable to call the onCompleted on our observer
  * This signals that all the saga observable streams have ended, which implies that the data fetchers have completed and all possible actions have been dispatched.
* Then the epic middleware is [restart](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/render-to-string.js#L25)
* And a final render pass is done to grab the hydrated app HTML string.

At this point the markup is sent to the user and it's then their responsibility to trigger observable clean up using the [epic's dispose method](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/create-epic.js#L83) which is a [composite disposable](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/create-epic.js#L43) of all the disposables created during the boot-up process.

I also provide a mirror of redux's `connect` which I called [`contain`](https://github.com/BerkeleyTrue/redux-epic/blob/master/src/contain.js#L38). This allows a user to declare which actions are responsible for triggering a specific data fetching saga and when the data is already present. This might be outside the scope of this project.

**Comment by steida** [comment]
Use something like https://github.com/este/este/blob/master/src/server/frontend/ServerFetchProvider.js

**Comment by BerkeleyTrue** [comment]
I've created a lib to try this out based on my work with React-Epic. Currently calling it React-Redux-Epic but will probably release it as React-Redux-Observable-SSR or RROSSR for short. 

I haven't had a chance to test it out, so beware of dragons. I'll be testing it out within the next two weeks.

My main concern ATM is that RO does not have some way of unsubscribing the internal observables it creates so their may be a memory leak. This will mean small changes to this project to expose some mechanism to unsubscribe/complete() the internal observables.


@jayphelps Thoughts?

**Comment by VasilioRuzanni** [comment]
@BerkeleyTrue What exactly do you refer to as "internal observables"?

**Comment by Pinpickle** [comment]
I ran into this problem as well, and found a solution that (so far) is working pretty well. It's quite similar to what was described by @BerkeleyTrue but in our app we have the extra complexity that actions are often triggered by new components being added.

For example, we load one resource (at which point the actions are supposedly done) which then renders a component which needs to load another resource (so the actions are not done!?). Because re-rendering isn't done automatically when using React SSR, this adds some thorns to the situation.

I came up with a strategy that works a bit like this:

1. Start counting actions
2. Render the app, let actions dispatch to the store
3. Close the stream of actions going to the epic (old epic) and divert them to a new epic
4. Wait for the old epic to complete (finish processing all actions)
5. Look at the action count - if it is zero, your state has all useful information and you can serve the rendered app. Otherwise, reset the counter and go back to 1

The nuts and bolts come down to two core components, a lifetime handler for epics (this is written in TypeScript):

```
/**
 * Composes multiple store epics together into a single root epic that we can track actions on.
 * This is achieved by separating the lifetime of our app's epic into distinct segments,
 * in which the amount of actions dispatched (by an epic or otherwise) are counted.
 *
 * These utilities are useful for server side rendering, so we can keep track of
 * asynchronous actions.
 *
 * - The counts are emitted via the count$ observable. This observable also tracks
 *   the lifetime of the root epic.
 * - The epic can be "segmented" via the segment() method
 * - The epic can be halted altogether via the halt() method
 *
 * @param epics - Individual epics to be composed together
 */
export function epicLifetimeHandler(epics: AppEpic<any>[]): EpicLifetime {
  const rawRootEpic = combineEpics(...epics);
  const command$ = new Subject<Command>();
  const actionCountPerSegment$ = new Subject<number>();

  const rootEpic: Epic<Action, any> = (action$, store) => {
    return action$
      .takeUntil(command$.filter(command => command === 'halt'))
      .window(command$.observeOn(asap).filter(command => command === 'segment'))
      .map(action$ => {
        let actionCount = 0;

        return rawRootEpic(
          ActionsObservable.from(action$.do(action => (actionCount += 1)).share()),
          store,
        ).do({ complete: () => actionCountPerSegment$.next(actionCount) });
      })
      .flatMap(outAction$ => outAction$)
      .do({
        complete: () => actionCountPerSegment$.complete(),
      });
  };

  return {
    rootEpic,
    actionCountPerSegment$: actionCountPerSegment$.asObservable(),
    segment: () => command$.next('segment'),
    halt: () => command$.next('halt'),
  };
}
```

And a sketch of rendering the app to completion looks a bit like this (just JS now):

```
function renderAppToCompletion(epicLifetime, appReactElement) {
  const appIsRenderedPromise = epicLifetime.actionCountPerSegment$.do(count => {
    if (count > 0) {
      // Actions were dispatched, there may be more work to do
      appReactElement.renderToString();
      epicLifetime.segment();
    } else {
      // No actions dispatched, we are done!
      // Halting lifetime will ultimately end the count observable and then the promise will resolve
      epicLifetime.halt();
    }
  }).toPromise();

  // Kick off with initial render
  appReactElement.renderToString();
  epicLifetime.segment();

  return appIsRenderedPromise;
}
```

**Comment by BerkeleyTrue** [comment]
> For example, we load one resource (at which point the actions are supposedly done) which then renders a component which needs to load another resource

@pinpickle sounds like this could be avoided by having the resource request automatically start the next request instead of the component being the trigger.

**Comment by BerkeleyTrue** [comment]
SSR using Redux-Observable is now possible with [react-redux-epic](https://github.com/BerkeleyTrue/react-redux-epic). If you have any questions, please feel free to post them here or in the Gitter chat room and I will happily help.

Happy Coding

**Comment by christophediprima** [comment]
Yes I have a question! Can you help me on making this work with found router ? Here is my post on the library:

https://github.com/4Catalyzer/found/issues/102

**Comment by christophediprima** [comment]
Ok guys! I managed to make it work ! And I am so happy! 😄

**Comment by gufy** [comment]
@christophediprima Hi, could you post your solution to the issue there please? Thanks.

**Comment by meinto** [comment]
I've written a library with redux-middleware to get redux-observables working with ssr...

https://github.com/meinto/redux-observables-server-side-rendering

I would be very happy to hear your opinions, suggestions and improvements about this approach.
I'm also thankful for every helpful pull request.

Basically it works like follows:

```js
// example epic
const myDataFetchingEpic = (action$, store, ssr) => actions$
  .ofType('SPECIFIC_ACTION_TYPE')
  .switchMap(action => {
    return ssr.observe(action, 
      getSomeDataFetchingObservable()
        .flatMap(response => [
          actionCreator1(response.data1),
          actionCreator2(response.data1),
        ])        
    )
  })
```

The redux middleware which is hooked into the redux-store does the rest...
The request waits until all triggered epics are processed and then sends the result to the Browser...

The setup is a bit expensive right now. But as i say. I'm happy about every contribution...
